content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+Thus, the originally well-defined program is transformed into one that is undefined.
+
+### Rewriting of Adjacent Data
+
+Given the following struct definition:
+
+```c
+struct {int a:17; int b:15} x;
+```
+
+A compiler may generate a data race from the following concurrent field assignments:
+
+```c
+x.a = 42; //thread1
+...
+x.b = 37; //thread2
+```
+
+At first glance, one might expect the above program to be race free, since both concurrent writes appear to access distinct variables.
+However, Pthreads does not define data races in terms of *program variables*, but rather in terms of **memory locations**.
+Since this term is not formally defined by Pthreads, compilers may have different program variables unexpectedly *share* a memory location.
+
+In the case above, a compiler targeting a 32-bit machine may choose to have the fields `x.a` and `x.b` share a memory location. This is probable in fact, since such machines are likely to have a 32-bit wide, rather than a 17 or 15-bit wide, store operation, necessitating the following implementation of bit-field assignment for `x.a = 42`:
+
+```c
+{
+  tmp = x;
+  tmp &= ~0x0001ffff;
+  tmp |= 42;
+  x = tmp;
+}
+```
+
+Thus, both fields `x.a` and `x.b` share memory location `x`, causing a data race when concurrently modified.
+
+As mentioned, this problematic transformation over adjacent bit-fields is motivated by architectural constraints. 
+However, a compiler can also deploy this transformation as a memory-saving optimization over adjacent fields.
+
+For example, given the following structure definition:
+```c
+struct { char a; char b; char c; char d; char e; char f; char g; char h; } x;
+```
+
+And a program that concurrently writes to adjacent fields:
+```c
+// Thread 1
+x.a = ‘a’ 
+
+ // Thread 2
+x.b = ‘b’; x.c=’c’; x.d=’d’; x.e=’e’; x.f=’f’; x.g=’g’; x.h=’h’;
+```
+
+A compiler that targets a 64-bit machine can optimize space usage by transforming Thread 2 as follows (assuming `x` is 64-bit aligned):
+```c
+x = ‘hgfedcb\0’ | x.a;
+```
+
+This optimization generates a data race, since both `x.a` and `x` share the same memory location and are concurrently modified.
+
+
+Since the writing of this paper, the C/C++ standard has been expanded to address this problem.
+Specifically, the C11 standard defines the memory locations of structure fields as follows:
+>A bit-field and an adjacent non-bit-field member are in **separate memory locations**. The same applies to two bit-fields, if one is declared inside a nested structure declaration and the other is not, or if the two are separated by a zero-length bit-field declaration, or if they are separated by a non-bit-field member declaration. It is **not safe** to concurrently update two non-atomic bit-fields in the same structure if all members declared between them are also (non-zero-length) bit-fields, no matter what the sizes of those intervening bit-fields happen to be.
```

Comment:
Since you're quoting from the C standard in this post, can you please link to the canonical source where you read this stuff at least once?